### PR TITLE
Add compile-time coverage for compiled code

### PR DESCRIPTION
**Pull request by sklam**
Let `Lower` gather coverage during compilation and write to active `Coverage` object.

It will use "numba_compiled" context to mark the lines visited by the compiler:

Screenshot 2024-03-20 at 3 33 02 PM

### CODE REVIEW

**Comment by tasansal** [comment]
Looking forward to this! Thanks for working on it!

**Comment by sklam** [comment]
Note that this is code coverage reported during compile-time, and not runtime code coverage, which I still cannot find a efficient way to do. 

> Will this make our numba code coverage on downstream libraries as well?

Yes, it works as long as the Numba can see an active `coverage.Coverage` object during compilation.

**Comment by sklam** [comment]
Mac failure is problem of the clang conda package: https://dev.azure.com/numba/numba/_build/results?buildId=17488&view=logs&j=7e7ebb60-d573-5e5d-b793-95b37f06385f&t=a4b2f274-a466-5129-a942-ad8f791ece9e&l=55

File: numba/misc/coverage_support.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    Make a singleton ``CoverageData``.
+    Avoid writing to disk. Other processes can corrupt the file.
+    """
+    covdata = coverage.CoverageData(no_disk=True)
+    cov = get_active_coverage()
+    assert cov is not None, "no active Coverage instance"
+
+    @atexit.register
+    def _finalize():
+        cov.get_data().update(covdata)
```
Comment (stuartarchibald) [inline numba/misc/coverage_support.py:54]: Note to self. This is the part that updates the existing active coverage with the information from the additional custom coverage object.

File: numba/misc/coverage_support.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    @abstractmethod
+    def close(self) -> None:
+        pass
+
+
+class NotifyCompilerCoverage(NotifyLocBase):
+    """
+    Use to notify ``coverage`` about compiled lines.
+
+    The compiled lines under "numba_compiled" context in the coverage data.
```
Comment (stuartarchibald) [inline numba/misc/coverage_support.py:75]:
```suggestion
    The compiled lines are under the "numba_compiled" context in the coverage
    data.
```

File: numba/misc/coverage_support.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    """
+    def __init__(self):
+        self._arcs_data = defaultdict(set)
+
+    def notify(self, loc: ir.Loc):
+        if loc.filename.endswith(".py"):
+            # The compiler doesn't actually know about arc.
+            self._arcs_data[loc.filename].add((loc.line, loc.line))
```
Comment (stuartarchibald) [inline numba/misc/coverage_support.py:83]: This is just creating an `arc` to itself, which I guess is good enough in terms of making it show up as being a "hit" line?

File: numba/tests/test_np_functions.py [hunk 4]
```diff
@@ -6276,7 +6276,7 @@ def foo():
         result, error = run_in_subprocess(code)
         # Assert that the bytestring "OK" was printed to stdout
         self.assertEqual(b"OK", result.strip())
-        self.assertEqual(b"", error.strip())
+        self.assertEqual(b"", error.strip(), msg=f"--ERROR--\n{error}\n")
```
Comment (stuartarchibald) [inline numba/tests/test_np_functions.py:6279]: Is this change needed for this PR?

**Review by stuartarchibald — COMMENTED** [review]
Thanks for working out how to implement this @sklam. I've reviewed the patch and it looks good, just a few minor things to take a look at. Thanks again!

File: buildscripts/incremental/setup_conda_environment.sh [hunk 5]
```diff
@@ -86,7 +86,7 @@ if [ "$TEST_SVML" == "yes" ]; then $CONDA_INSTALL -c numba icc_rt; fi
 # Install Intel TBB parallel backend
 if [ "$TEST_THREADING" == "tbb" ]; then $CONDA_INSTALL "tbb>=2021.6" "tbb-devel>=2021.6"; fi
 # Install typeguard
-if [ "$RUN_TYPEGUARD" == "yes" ]; then $CONDA_INSTALL "conda-forge::typeguard==3.0.1"; fi
+if [ "$RUN_TYPEGUARD" == "yes" ]; then $CONDA_INSTALL typeguard; fi
```
[the remark this answers is not in the document]
Comment (sklam) [reply buildscripts/incremental/setup_conda_environment.sh:89]: It is necessary for a getting newer versions of it. The version that the script pinned to has trouble with integrating with coverage.

File: numba/misc/coverage_support.py (hunk 3 above)
Comment (sklam) [reply numba/misc/coverage_support.py:83]: Branch-coverage enabled runs requires the use of arc data.  If we use line data, it cannot be merged with branch-coverage enabled data. But arc data can be merged with non-branch-coverage enabled data.

File: numba/tests/test_np_functions.py (hunk 4 above)
Comment (sklam) [reply numba/tests/test_np_functions.py:6279]: No, just good for debugging.

File: buildscripts/incremental/setup_conda_environment.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply buildscripts/incremental/setup_conda_environment.sh:89]: ah ok, thanks for explaining.

File: numba/tests/test_np_functions.py (hunk 4 above)
Comment (stuartarchibald) [reply numba/tests/test_np_functions.py:6279]: OK, am happy to leave it.

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes.

**Comment by stuartarchibald** [comment]
I am just going to close and reopen this to trigger a rebuild on RTD now that #9675 resolved the issue.

**Comment by stuartarchibald** [comment]
@sklam I think that this branch is going to need to have `numba:main` merged in to pick up the changes in #9675 so that RTD will build.

**Comment by sklam** [comment]
RTD is happy now

**Review by stuartarchibald — APPROVED** [review]
Thanks for fixing the RTD issue.
